specs/language/resources.tex

```diff
@@ -663,19 +664,237 @@
+\textit{namespace-declaration} or a \textit{variable-declaration} is going to be
+processed as if it was declared outside of the group at the
+global scope.
+
+\p Nesting of \texttt{cbuffer} declaration groups is not allowed.
+
+\p Any \textit{namespace-declaration} in the \texttt{cbuffer} declaration group can contain only declarations allowed by \textit{cbuffer-declaration}.
+
+\p For example:
+
+\begin{HLSL}
+  cbuffer MyConstants {
+    float4 CameraPos;
+  };
+
+  float4 getCameraPosition() {
+    return CameraPos;
+  }
+\end{HLSL}
+
+\Sub{Constant Buffer Class}{Resources.cnbuf.cbclass}
+
+\p Another way of declaring constant buffers is by using the
+\texttt{ConstantBuffer<T>} resource class.
+
+\p The template parameter \texttt{T} must be a class type (\ref{Classes}). 
+
+\begin{HLSL}
+  template <typename T>
+   class ConstantBuffer {
+   public:
+     ConstantBuffer();
+     ConstantBuffer(const ConstantBuffer &buf);
+     ConstantBuffer &operator=(ConstantBuffer &buf);
+  };
+\end{HLSL}
+
+\p The \textit{shader constants} are the fields of the class type \texttt{T}.
+They can be referenced from anywhere within the translation unit after the
+\texttt{ConstantBuffer<T>} declaration as if they were fields of the
+\texttt{ConstantBuffer<T>} class.
+
+\p For example:
+
+\begin{HLSL}
+  struct MyConstants {
+    float4 CameraPos;
+  };
+
+  ConstantBuffer<MyConstants> CB;
+
+  float4 getCameraPosition() {
+    return CB.CameraPos;
+  }
+\end{HLSL}
+
+\p The layout rules for constant buffer declared with the
+\texttt{ConstantBuffer<T>} syntax are the same as for named \texttt{cbuffer}
+declaration groups.
+
+\p \texttt{ConstantBuffer<T>} can be defined at local scope to represent a local
+reference to a constant buffer that can be associated with a global buffer when
+assigned. The reference must be resolvable to a unique global buffer declaration
+before use.
+
+\p Constant buffers cannot be implicitly nor explicitly cast from one type to
+another. This means that local \texttt{ConstantBuffer} can only be assigned from
+\texttt{ConstantBuffer} with the same template type \texttt{T}.
+
+\Sub{Constant Buffer Layout}{Resources.cnbuf.lay}
+
+\p Constant buffer is arranged like an array of 16-byte rows, or 4-component
+vectors of 32-bit elements. Shader constants are arranged into the buffer in the
+order they were declared based on following rules.
+
+\p Specific basic types are packed into the last used row of the buffer if they
+fit into the available space, starting at the next available aligned position.
+These types include:
+\begin{itemize}
+\item \textit{scalar types}
+\item \textit{vector types}
+\item \textit{matrix types} with only one row in \textit{column\_major} storage layout
+\end{itemize}
+
+\p If they do not fit the remaining space of the row, they will placed at the
+start a new 16-byte aligned row.
+
+\p Vectors are aligned by the size of a single vector element type.
```

Comment:
This doesn't account for the row wrapping right? Shouldn't it be something like:

`Vectors are aligned by the size of a single vector element type unless that alignment results in crossing the 16 byte row boundary, in which case it is aligned to the next row"